<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="author" content="Natalie Elphick" />
  <title>C3 R Bootcamp Bonus Day</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="C3_bonus_day_files/reveal.js-3.3.0.1/css/reveal.css"/>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="C3_bonus_day_files/reveal.js-3.3.0.1/css/theme/simple.css" id="theme">


  <!-- some tweaks to reveal css -->
  <style type="text/css">
    .reveal h1 { font-size: 2.0em; }
    .reveal h2 { font-size: 1.5em;  }
    .reveal h3 { font-size: 1.25em;	}
    .reveal h4 { font-size: 1em;	}

    .reveal .slides>section,
    .reveal .slides>section>section {
      padding: 0px 0px;
    }



    .reveal table {
      border-width: 1px;
      border-spacing: 2px;
      border-style: dotted;
      border-color: gray;
      border-collapse: collapse;
      font-size: 0.7em;
    }

    .reveal table th {
      border-width: 1px;
      padding-left: 10px;
      padding-right: 25px;
      font-weight: bold;
      border-style: dotted;
      border-color: gray;
    }

    .reveal table td {
      border-width: 1px;
      padding-left: 10px;
      padding-right: 25px;
      border-style: dotted;
      border-color: gray;
    }


  </style>

    <style type="text/css">code{white-space: pre;}</style>

    <link rel="stylesheet" href="style.css"/>

<!-- Printing and PDF exports -->
<script id="paper-css" type="application/dynamic-css">

/* Default Print Stylesheet Template
   by Rob Glazebrook of CSSnewbie.com
   Last Updated: June 4, 2008

   Feel free (nay, compelled) to edit, append, and
   manipulate this file as you see fit. */


@media print {

	/* SECTION 1: Set default width, margin, float, and
	   background. This prevents elements from extending
	   beyond the edge of the printed page, and prevents
	   unnecessary background images from printing */
	html {
		background: #fff;
		width: auto;
		height: auto;
		overflow: visible;
	}
	body {
		background: #fff;
		font-size: 20pt;
		width: auto;
		height: auto;
		border: 0;
		margin: 0 5%;
		padding: 0;
		overflow: visible;
		float: none !important;
	}

	/* SECTION 2: Remove any elements not needed in print.
	   This would include navigation, ads, sidebars, etc. */
	.nestedarrow,
	.controls,
	.fork-reveal,
	.share-reveal,
	.state-background,
	.reveal .progress,
	.reveal .backgrounds {
		display: none !important;
	}

	/* SECTION 3: Set body font face, size, and color.
	   Consider using a serif font for readability. */
	body, p, td, li, div {
		font-size: 20pt!important;
		font-family: Georgia, "Times New Roman", Times, serif !important;
		color: #000;
	}

	/* SECTION 4: Set heading font face, sizes, and color.
	   Differentiate your headings from your body text.
	   Perhaps use a large sans-serif for distinction. */
	h1,h2,h3,h4,h5,h6 {
		color: #000!important;
		height: auto;
		line-height: normal;
		font-family: Georgia, "Times New Roman", Times, serif !important;
		text-shadow: 0 0 0 #000 !important;
		text-align: left;
		letter-spacing: normal;
	}
	/* Need to reduce the size of the fonts for printing */
	h1 { font-size: 28pt !important;  }
	h2 { font-size: 24pt !important; }
	h3 { font-size: 22pt !important; }
	h4 { font-size: 22pt !important; font-variant: small-caps; }
	h5 { font-size: 21pt !important; }
	h6 { font-size: 20pt !important; font-style: italic; }

	/* SECTION 5: Make hyperlinks more usable.
	   Ensure links are underlined, and consider appending
	   the URL to the end of the link for usability. */
	a:link,
	a:visited {
		color: #000 !important;
		font-weight: bold;
		text-decoration: underline;
	}
	/*
	.reveal a:link:after,
	.reveal a:visited:after {
		content: " (" attr(href) ") ";
		color: #222 !important;
		font-size: 90%;
	}
	*/


	/* SECTION 6: more reveal.js specific additions by @skypanther */
	ul, ol, div, p {
		visibility: visible;
		position: static;
		width: auto;
		height: auto;
		display: block;
		overflow: visible;
		margin: 0;
		text-align: left !important;
	}
	.reveal pre,
	.reveal table {
		margin-left: 0;
		margin-right: 0;
	}
	.reveal pre code {
		padding: 20px;
		border: 1px solid #ddd;
	}
	.reveal blockquote {
		margin: 20px 0;
	}
	.reveal .slides {
		position: static !important;
		width: auto !important;
		height: auto !important;

		left: 0 !important;
		top: 0 !important;
		margin-left: 0 !important;
		margin-top: 0 !important;
		padding: 0 !important;
		zoom: 1 !important;

		overflow: visible !important;
		display: block !important;

		text-align: left !important;
		-webkit-perspective: none;
		   -moz-perspective: none;
		    -ms-perspective: none;
		        perspective: none;

		-webkit-perspective-origin: 50% 50%;
		   -moz-perspective-origin: 50% 50%;
		    -ms-perspective-origin: 50% 50%;
		        perspective-origin: 50% 50%;
	}
	.reveal .slides section {
		visibility: visible !important;
		position: static !important;
		width: auto !important;
		height: auto !important;
		display: block !important;
		overflow: visible !important;

		left: 0 !important;
		top: 0 !important;
		margin-left: 0 !important;
		margin-top: 0 !important;
		padding: 60px 20px !important;
		z-index: auto !important;

		opacity: 1 !important;

		page-break-after: always !important;

		-webkit-transform-style: flat !important;
		   -moz-transform-style: flat !important;
		    -ms-transform-style: flat !important;
		        transform-style: flat !important;

		-webkit-transform: none !important;
		   -moz-transform: none !important;
		    -ms-transform: none !important;
		        transform: none !important;

		-webkit-transition: none !important;
		   -moz-transition: none !important;
		    -ms-transition: none !important;
		        transition: none !important;
	}
	.reveal .slides section.stack {
		padding: 0 !important;
	}
	.reveal section:last-of-type {
		page-break-after: avoid !important;
	}
	.reveal section .fragment {
		opacity: 1 !important;
		visibility: visible !important;

		-webkit-transform: none !important;
		   -moz-transform: none !important;
		    -ms-transform: none !important;
		        transform: none !important;
	}
	.reveal section img {
		display: block;
		margin: 15px 0px;
		background: rgba(255,255,255,1);
		border: 1px solid #666;
		box-shadow: none;
	}

	.reveal section small {
		font-size: 0.8em;
	}

}  
</script>


<script id="pdf-css" type="application/dynamic-css">
    
/**
 * This stylesheet is used to print reveal.js
 * presentations to PDF.
 *
 * https://github.com/hakimel/reveal.js#pdf-export
 */

* {
	-webkit-print-color-adjust: exact;
}

body {
	margin: 0 auto !important;
	border: 0;
	padding: 0;
	float: none !important;
	overflow: visible;
}

html {
	width: 100%;
	height: 100%;
	overflow: visible;
}

/* Remove any elements not needed in print. */
.nestedarrow,
.reveal .controls,
.reveal .progress,
.reveal .playback,
.reveal.overview,
.fork-reveal,
.share-reveal,
.state-background {
	display: none !important;
}

h1, h2, h3, h4, h5, h6 {
	text-shadow: 0 0 0 #000 !important;
}

.reveal pre code {
	overflow: hidden !important;
	font-family: Courier, 'Courier New', monospace !important;
}

ul, ol, div, p {
	visibility: visible;
	position: static;
	width: auto;
	height: auto;
	display: block;
	overflow: visible;
	margin: auto;
}
.reveal {
	width: auto !important;
	height: auto !important;
	overflow: hidden !important;
}
.reveal .slides {
	position: static;
	width: 100%;
	height: auto;

	left: auto;
	top: auto;
	margin: 0 !important;
	padding: 0 !important;

	overflow: visible;
	display: block;

	-webkit-perspective: none;
	   -moz-perspective: none;
	    -ms-perspective: none;
	        perspective: none;

	-webkit-perspective-origin: 50% 50%; /* there isn't a none/auto value but 50-50 is the default */
	   -moz-perspective-origin: 50% 50%;
	    -ms-perspective-origin: 50% 50%;
	        perspective-origin: 50% 50%;
}

.reveal .slides section {
	page-break-after: always !important;

	visibility: visible !important;
	position: relative !important;
	display: block !important;
	position: relative !important;

	margin: 0 !important;
	padding: 0 !important;
	box-sizing: border-box !important;
	min-height: 1px;

	opacity: 1 !important;

	-webkit-transform-style: flat !important;
	   -moz-transform-style: flat !important;
	    -ms-transform-style: flat !important;
	        transform-style: flat !important;

	-webkit-transform: none !important;
	   -moz-transform: none !important;
	    -ms-transform: none !important;
	        transform: none !important;
}

.reveal section.stack {
	margin: 0 !important;
	padding: 0 !important;
	page-break-after: avoid !important;
	height: auto !important;
	min-height: auto !important;
}

.reveal img {
	box-shadow: none;
}

.reveal .roll {
	overflow: visible;
	line-height: 1em;
}

/* Slide backgrounds are placed inside of their slide when exporting to PDF */
.reveal section .slide-background {
	display: block !important;
	position: absolute;
	top: 0;
	left: 0;
	width: 100%;
	z-index: -1;
}

/* All elements should be above the slide-background */
.reveal section>* {
	position: relative;
	z-index: 1;
}

/* Display slide speaker notes when 'showNotes' is enabled */
.reveal .speaker-notes-pdf {
	display: block;
	width: 100%;
	max-height: none;
	left: auto;
	top: auto;
	z-index: 100;
}

/* Display slide numbers when 'slideNumber' is enabled */
.reveal .slide-number-pdf {
	display: block;
	position: absolute;
	font-size: 14px;
}

</script>


<script>
var style = document.createElement( 'style' );
style.type = 'text/css';
var style_script_id = window.location.search.match( /print-pdf/gi ) ? 'pdf-css' : 'paper-css';
var style_script = document.getElementById(style_script_id).text;
style.innerHTML = style_script;
document.getElementsByTagName('head')[0].appendChild(style);
</script>

    <script src="C3_bonus_day_files/header-attrs-2.25/header-attrs.js"></script>
    <link href="C3_bonus_day_files/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
    <link href="C3_bonus_day_files/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
</head>
<body>
  <div class="reveal">
    <div class="slides">

<section>
    <h1 class="title">C3 R Bootcamp Bonus Day</h1>
    <h2 class="author">Natalie Elphick</h2>
    <h3 class="date">November 9th</h3>
</section>

<section id="section" class="slide level2">
<h2></h2>
<center>
<em>Press the ? key for tips on navigating these slides</em>
</center>
</section>
<section id="topics" class="slide level2">
<h2>Topics</h2>
<ul>
<li><p>RMD notebooks, R projects, and R environments</p></li>
<li><p><code>apply</code> and <code>map</code> functions</p></li>
<li><p>Intro to Linear Models</p></li>
<li><p>How to find bioinformatics packages for your use case</p></li>
<li><p>ChatGPT Tips for Bioinformatics</p></li>
<li><p>Additional Resources</p></li>
</ul>
</section>
<section>
<section id="rmd-notebooks-r-projects-and-r-environments"
class="title-slide slide level1">
<h1>RMD notebooks, R projects, and R environments</h1>

</section>
<section id="r-markdown" class="slide level2">
<h2>R Markdown</h2>
<ul>
<li>A file format combining <code>R</code> code with <a
href="https://www.markdownguide.org/basic-syntax/">Markdown</a> for text
formatting.</li>
<li>Designed for creating reproducible research reports in various
formats (HTML, PDF, Word).</li>
<li>Let’s create an <code>Rmd</code> file in <code>RStudio</code> to
explore the basics of how they work:</li>
<li>File -&gt; New File -&gt; R Markdown</li>
</ul>
</section>
<section id="r-markdown-advanced-usage" class="slide level2">
<h2>R Markdown Advanced Usage</h2>
<ul>
<li><strong>Presentations:</strong> Creating slides (like these) with <a
href="https://github.com/rstudio/revealjs">revealjs</a>.</li>
<li><strong>Publications:</strong> Authoring online books that combine
narrative, code, and output with <a
href="https://bookdown.org/">bookdown</a>.</li>
<li><strong>Interactive Documents:</strong> Developing interactive
tutorials or dashboards with <a
href="https://rstudio.github.io/learnr/">learnR</a> and other embedded
applications.</li>
</ul>
</section>
<section id="r-projects" class="slide level2">
<h2>R Projects</h2>
<ul>
<li>A project is just a working directory designated with a
<code>.RProj</code></li>
<li>Opening the <code>.RProj</code> file sets the working directory to
the location of the <code>.RProj</code> file</li>
<li>Helps implement a common best practice for data analysis
<ul>
<li>Having the input data, R scripts, results and figures all under one
directory</li>
</ul></li>
<li>See <a
href="https://r4ds.had.co.nz/workflow-projects.html#workflow-projects">this
chapter</a> for more information</li>
</ul>
</section>
<section id="r-environments" class="slide level2">
<h2>R Environments</h2>
<ul>
<li>At a basic level the environment is the data structure used to
<strong>scope</strong> variables and functions.</li>
<li><strong>Scoping:</strong> the concept of defining the visibility and
accessibility of variables/functions within a program.</li>
<li>Similar to a named <code>list</code>, except:
<ul>
<li>Every name must be unique.</li>
<li>The names in an environment are not ordered.</li>
</ul></li>
</ul>
</section></section>
<section>
<section id="apply-and-map-functions" class="title-slide slide level1">
<h1>Apply and Map Functions</h1>

</section>
<section id="apply" class="slide level2">
<h2>apply()</h2>
<ul>
<li><p>Apply a function over a margin (1 = rows, 2 = columns) and return
the results in an array</p></li>
<li><p>Alternative to for loops</p></li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">apply</span>(<span class="at">X =</span> cars, <span class="at">MARGIN =</span> <span class="dv">2</span>, <span class="at">FUN =</span>  mean)</span></code></pre></div>
<pre><code>## speed  dist 
## 15.40 42.98</code></pre>
<ul>
<li><p><code>apply()</code> works for matrices and data frames, there is
also:</p>
<ul>
<li><p><code>lapply()</code> for lists</p></li>
<li><p><code>vapply()</code> for vectors</p></li>
</ul></li>
</ul>
</section>
<section id="apply-custom-functions" class="slide level2">
<h2>apply() : custom functions</h2>
<ul>
<li>You can supply a custom function within apply</li>
<li>These nameless functions are known as anonymous functions or lambda
expressions</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">apply</span>(<span class="at">X =</span> cars, <span class="at">MARGIN =</span> <span class="dv">2</span>,</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">FUN =</span>  <span class="cf">function</span>(col) {</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>        col_mean <span class="ot">&lt;-</span> <span class="fu">sum</span>(col) <span class="sc">/</span> <span class="fu">length</span>(col)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">return</span>(col_mean)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>})</span></code></pre></div>
<pre><code>## speed  dist 
## 15.40 42.98</code></pre>
<ul>
<li>The above is much faster and more memory efficient than looping</li>
</ul>
</section>
<section id="apply-converting-for-loops-to-apply" class="slide level2">
<h2>apply() : Converting for loops to apply</h2>
<ul>
<li>Since converting syntax can be tedious this is a good use case for
ChatGPT</li>
<li>open the file</li>
</ul>
</section>
<section id="purrmap-tidyverse-apply" class="slide level2">
<h2>purr::map() : tidyverse apply</h2>
<ul>
<li><a href="https://purrr.tidyverse.org/reference/map.html">map()</a>
functions are used to apply a function to each element of a list or
vector</li>
<li>They are type-stable, meaning they always return the same type of
output (unlike apply)</li>
<li><code>map_lgl()</code>, <code>map_int()</code>,
<code>map_dbl()</code> and <code>map_chr()</code> return a vector of the
indicated type</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(purrr)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">map</span>(cars, <span class="sc">~</span> <span class="fu">mean</span>(.x)) </span></code></pre></div>
<pre><code>## $speed
## [1] 15.4
## 
## $dist
## [1] 42.98</code></pre>
</section>
<section id="purrmap-cont." class="slide level2">
<h2>purr::map() (cont.)</h2>
<ul>
<li>Chaining multiple map calls can be a powerful tool to get complex
outputs from a few lines of code</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>mtcars <span class="sc">%&gt;%</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">split</span>(mtcars<span class="sc">$</span>cyl) <span class="sc">%&gt;%</span> </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">map</span>(<span class="at">.f =</span> <span class="cf">function</span>(df) <span class="fu">lm</span>(mpg <span class="sc">~</span> wt, <span class="at">data =</span> df)) <span class="sc">%&gt;%</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">map</span>(summary) <span class="sc">%&gt;%</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">map_dbl</span>(<span class="st">&quot;r.squared&quot;</span>)</span></code></pre></div>
<pre><code>##         4         6         8 
## 0.5086326 0.4645102 0.4229655</code></pre>
</section></section>
<section>
<section id="intro-to-linear-models-in-r"
class="title-slide slide level1">
<h1>Intro to Linear Models in R</h1>

</section>
<section id="common-applications" class="slide level2">
<h2>Common Applications</h2>
<ul>
<li>RNA-Seq differential expression analyses</li>
<li>GWAS (for continuous traits)</li>
<li>eQTL analyses</li>
<li>Microarray data analyses</li>
</ul>
</section>
<section id="linear-model-background" class="slide level2">
<h2>Linear Model Background</h2>
<p><span class="math display">\[
Y \sim \beta_0 + x_1\beta_1 + x_2\beta_2 ...x_n\beta_n + \epsilon
\]</span></p>
<ul>
<li><p><span class="math inline">\(Y\)</span> is our outcome</p></li>
<li><p>The covariates <span class="math inline">\(x\)</span> can be</p>
<ul>
<li><p>Continuous (age, weight, temperature, etc.)</p></li>
<li><p>Categorical (treatment, batch, genotype, etc.)</p></li>
</ul></li>
<li><p>The <span class="math inline">\(\beta\)</span>’s are unknown
parameters to be estimated.</p></li>
<li><p>The error term <span class="math inline">\(\epsilon\)</span> is
assumed to be normally distributed with a variance that is constant
across the range of the data.</p></li>
</ul>
</section>
<section id="lm-r-function-for-linear-models" class="slide level2">
<h2>lm() : R Function for Linear Models</h2>
<ul>
<li><code>lm()</code> expects a model formula</li>
<li><code>lm()</code> does more than just linear regression</li>
<li>The formula is specified in the following format:</li>
</ul>
<pre><code>Y ~ X
gene ~ age               # Regression
gene ~ sex               # One-way ANOVA
gene ~ age + genotype    # Analysis of covariance</code></pre>
<ul>
<li>See <a
href="https://www.ucd.ie/ecomodel/Resources/R_formulae_WebVersion.html">Formulas
in R</a> for an in depth explanation of using formulas to specify models
in R</li>
</ul>
</section>
<section id="example-dataset" class="slide level2">
<h2>Example Dataset</h2>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;https://raw.githubusercontent.com/ucdavis-bioinformatics-training/2018-September-Bioinformatics-Prerequisites/master/friday/lm_example_data.csv&quot;</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dat) <span class="sc">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  kableExtra<span class="sc">::</span><span class="fu">kable</span>()</span></code></pre></div>
<table>
<thead>
<tr>
<th style="text-align:right;">
sample
</th>
<th style="text-align:right;">
expression
</th>
<th style="text-align:left;">
batch
</th>
<th style="text-align:left;">
treatment
</th>
<th style="text-align:left;">
time
</th>
<th style="text-align:right;">
temperature
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1.2139625
</td>
<td style="text-align:left;">
Batch1
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
time1
</td>
<td style="text-align:right;">
11.76575
</td>
</tr>
<tr>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
1.4796581
</td>
<td style="text-align:left;">
Batch1
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
time2
</td>
<td style="text-align:right;">
12.16330
</td>
</tr>
<tr>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
1.0878287
</td>
<td style="text-align:left;">
Batch1
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
time1
</td>
<td style="text-align:right;">
10.54195
</td>
</tr>
<tr>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
1.4438585
</td>
<td style="text-align:left;">
Batch1
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
time2
</td>
<td style="text-align:right;">
10.07642
</td>
</tr>
<tr>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
0.6371621
</td>
<td style="text-align:left;">
Batch1
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
time1
</td>
<td style="text-align:right;">
12.03721
</td>
</tr>
<tr>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
2.1226740
</td>
<td style="text-align:left;">
Batch1
</td>
<td style="text-align:left;">
B
</td>
<td style="text-align:left;">
time2
</td>
<td style="text-align:right;">
13.49573
</td>
</tr>
</tbody>
</table>
</section>
<section id="fitting-the-model" class="slide level2">
<h2>Fitting the Model</h2>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>oneway_model <span class="ot">&lt;-</span> <span class="fu">lm</span>(expression <span class="sc">~</span> treatment, <span class="at">data =</span> dat)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(oneway_model)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = expression ~ treatment, data = dat)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -3.9310 -0.5353  0.1790  0.7725  3.6114 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)   1.1725     0.7783   1.506    0.148    
## treatmentB    0.4455     1.1007   0.405    0.690    
## treatmentC    0.9028     1.1007   0.820    0.422    
## treatmentD    2.5537     1.1007   2.320    0.031 *  
## treatmentE    7.4140     1.1007   6.735 1.49e-06 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 1.74 on 20 degrees of freedom
## Multiple R-squared:  0.7528, Adjusted R-squared:  0.7033 
## F-statistic: 15.22 on 4 and 20 DF,  p-value: 7.275e-06</code></pre>
</section>
<section id="understanding-the-model-outputs" class="slide level2">
<h2>Understanding the Model Outputs</h2>
<ul>
<li>“Coefficients” refer to the <span
class="math inline">\(\beta\)</span>’s</li>
<li>“Estimate” is the estimate of each coefficient (effect size)</li>
<li>“Std. Error” is the standard error of the estimate</li>
<li>“t value” is the coefficient divided by its standard error</li>
<li>“Pr(&gt;|t|)”is the p-value for the coefficient</li>
</ul>
</section>
<section id="understanding-the-model-outputs-cont."
class="slide level2">
<h2>Understanding the Model Outputs (cont.)</h2>
<ul>
<li>The residual standard error is the estimate of the variance of
ϵ</li>
<li>Degrees of freedom is the sample size minus # of coefficients
estimated</li>
<li>R-squared is (roughly) the proportion of variance in the outcome
explained by the model</li>
<li>The F-statistic compares the fit of the model as a whole to the null
model (with no covariates)</li>
</ul>
</section>
<section id="what-do-the-model-coefficients-mean" class="slide level2">
<h2>What do the Model Coefficients Mean?</h2>
<ul>
<li>R uses reference group coding or “treatment contrasts”.</li>
<li>For categorical covariates, the first level alphabetically (or first
factor level) is treated as the reference group.</li>
<li>The reference group is represented by the intercept.</li>
<li>Using control as a reference level is a common practice.</li>
</ul>
</section>
<section id="coefficients-in-the-example" class="slide level2">
<h2>Coefficients in the Example</h2>
<ul>
<li><p>Coefficients for other groups are the difference from the
reference:</p>
<p>(Intercept) = the mean for treatmentA<br />
treatmentB = the mean for treatmentB - the mean for treatmentA<br />
treatmentC = the mean for treatmentC - the mean for treatmentA<br />
etc.</p></li>
</ul>
</section>
<section id="broomtidy-summary-table-of-model-fit" class="slide level2">
<h2>broom::tidy : Summary Table of Model Fit</h2>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>broom<span class="sc">::</span><span class="fu">tidy</span>(oneway_model) <span class="sc">%&gt;%</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(p.value)) <span class="sc">%&gt;%</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  kableExtra<span class="sc">::</span><span class="fu">kable</span>()</span></code></pre></div>
<table>
<thead>
<tr>
<th style="text-align:left;">
term
</th>
<th style="text-align:right;">
estimate
</th>
<th style="text-align:right;">
std.error
</th>
<th style="text-align:right;">
statistic
</th>
<th style="text-align:right;">
p.value
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
treatmentB
</td>
<td style="text-align:right;">
0.4455249
</td>
<td style="text-align:right;">
1.1007321
</td>
<td style="text-align:right;">
0.4047532
</td>
<td style="text-align:right;">
0.6899552
</td>
</tr>
<tr>
<td style="text-align:left;">
treatmentC
</td>
<td style="text-align:right;">
0.9027755
</td>
<td style="text-align:right;">
1.1007321
</td>
<td style="text-align:right;">
0.8201592
</td>
<td style="text-align:right;">
0.4217930
</td>
</tr>
<tr>
<td style="text-align:left;">
(Intercept)
</td>
<td style="text-align:right;">
1.1724940
</td>
<td style="text-align:right;">
0.7783351
</td>
<td style="text-align:right;">
1.5064128
</td>
<td style="text-align:right;">
0.1475935
</td>
</tr>
<tr>
<td style="text-align:left;">
treatmentD
</td>
<td style="text-align:right;">
2.5536669
</td>
<td style="text-align:right;">
1.1007321
</td>
<td style="text-align:right;">
2.3199713
</td>
<td style="text-align:right;">
0.0310337
</td>
</tr>
<tr>
<td style="text-align:left;">
treatmentE
</td>
<td style="text-align:right;">
7.4139642
</td>
<td style="text-align:right;">
1.1007321
</td>
<td style="text-align:right;">
6.7354847
</td>
<td style="text-align:right;">
0.0000015
</td>
</tr>
</tbody>
</table>
</section>
<section id="model-matrices" class="slide level2">
<h2>Model Matrices</h2>
<ul>
<li><p>Tools like <code>limma</code> and <code>edgeR</code> use
<em>design matrices</em> to specify the model</p></li>
<li><p>The design matrix <span class="math inline">\(X\)</span></p>
<ul>
<li><p>One row for each observation (sample)</p></li>
<li><p>One column for each model coefficient (treatment)</p></li>
</ul></li>
</ul>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> treatment, <span class="at">data =</span> dat)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(X,<span class="dv">4</span>)</span></code></pre></div>
<pre><code>##   (Intercept) treatmentB treatmentC treatmentD treatmentE
## 1           1          0          0          0          0
## 2           1          0          0          0          0
## 3           1          0          0          0          0
## 4           1          0          0          0          0</code></pre>
</section>
<section id="understanding-model-matrices" class="slide level2">
<h2>Understanding Model Matrices</h2>
<ul>
<li>Linear regression case:</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(expression <span class="sc">~</span> temperature, <span class="at">data =</span> dat)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>X <span class="sc">%&gt;%</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>() <span class="sc">%&gt;%</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  kableExtra<span class="sc">::</span><span class="fu">kable</span>()</span></code></pre></div>
<table>
<thead>
<tr>
<th style="text-align:right;">
(Intercept)
</th>
<th style="text-align:right;">
temperature
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
11.76575
</td>
</tr>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
12.16330
</td>
</tr>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
10.54195
</td>
</tr>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
10.07642
</td>
</tr>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
12.03721
</td>
</tr>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
13.49573
</td>
</tr>
</tbody>
</table>
</section>
<section id="understanding-model-matrices-cont." class="slide level2">
<h2>Understanding Model Matrices (cont.)</h2>
<ul>
<li>One-way ANOVA:</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(expression <span class="sc">~</span> treatment, <span class="at">data =</span> dat)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>X <span class="sc">%&gt;%</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(treatmentB)) </span></code></pre></div>
<pre><code>##    (Intercept) treatmentB treatmentC treatmentD treatmentE
## 6            1          1          0          0          0
## 7            1          1          0          0          0
## 8            1          1          0          0          0
## 9            1          1          0          0          0
## 10           1          1          0          0          0
## 1            1          0          0          0          0
## 2            1          0          0          0          0
## 3            1          0          0          0          0
## 4            1          0          0          0          0
## 5            1          0          0          0          0
## 11           1          0          1          0          0
## 12           1          0          1          0          0
## 13           1          0          1          0          0
## 14           1          0          1          0          0
## 15           1          0          1          0          0
## 16           1          0          0          1          0
## 17           1          0          0          1          0
## 18           1          0          0          1          0
## 19           1          0          0          1          0
## 20           1          0          0          1          0
## 21           1          0          0          0          1
## 22           1          0          0          0          1
## 23           1          0          0          0          1
## 24           1          0          0          0          1
## 25           1          0          0          0          1</code></pre>
</section>
<section id="further-learning" class="slide level2">
<h2>Further Learning</h2>
<ul>
<li>Follow the <a
href="https://ucdavis-bioinformatics-training.github.io/2019-March-Bioinformatics-Prerequisites/thursday/linear_models.html">full
tutorial</a> and complete the exercises</li>
<li>Learn about more advanced implementations of linear models:<br />
</li>
<li><a
href="https://bookdown.org/steve_midway/DAR/glms-generalized-linear-models.html">Generalized
Linear Models</a><br />
</li>
<li><a
href="https://bookdown.org/steve_midway/DAR/random-effects.html">Random
Effects</a></li>
</ul>
</section></section>
<section>
<section id="chatgpt-tips-for-bioinformatics"
class="title-slide slide level1">
<h1>ChatGPT Tips for Bioinformatics</h1>

</section>
<section id="general-tips" class="slide level2">
<h2>General Tips</h2>
<ul>
<li>Always confirm ChatGPT’s outputs are correct</li>
<li>Provide as much detail as possible about the problem</li>
<li>Use separate chats for separate tasks/projects</li>
<li>Try the ‘Custom Instructions’ function that adds additional
information to every query you run</li>
<li>The new release can visit webpages (GPT 4 only), which can help get
more specific answers</li>
</ul>
</section>
<section id="code-tips" class="slide level2">
<h2>Code Tips</h2>
<ul>
<li>Commented R code yields better responses in my experience</li>
<li>Provide the code and error message in the same query</li>
<li>ChatGPT works well to convert syntax and improve your code:
<ul>
<li>“Turn this loop into a function : [your code]”</li>
<li>“Is there a better way to do this : [your code]”</li>
</ul></li>
<li>Check out the file: <code>example_code/1_for_loop_to_apply.R</code>
for an example use case</li>
</ul>
</section></section>
<section>
<section id="finding-bioinformatics-packages"
class="title-slide slide level1">
<h1>Finding Bioinformatics Packages</h1>

</section>
<section id="key-questions" class="slide level2">
<h2>Key Questions</h2>
<ul>
<li>What assay was the package designed for?</li>
<li>When was the last release?</li>
<li>Is it maintained (frequent updates)?</li>
<li>Does it work on all operating systems?</li>
<li>Are other people using it?</li>
<li>Do they respond to github issues?</li>
<li>Is there a benchmarking paper?</li>
</ul>
</section>
<section id="bioconductor-and-cran" class="slide level2">
<h2>BioConductor and CRAN</h2>
<ul>
<li><p>Both of these have stringent requirements for packages they host
(eg. for BioConductor they have to run on all major operating
systems)</p></li>
<li><p>Prefer BioConductor packages if available over CRAN</p></li>
<li><p>Prefer CRAN packages over ones only hosted on GitHub</p></li>
</ul>
</section>
<section id="start-with-the-assay" class="slide level2">
<h2>Start with the Assay</h2>
<ul>
<li>Click <a
href="https://www.bioconductor.org/packages/release/BiocViews.html#___Sequencing">here</a>
to go to BioC views</li>
<li>Pick the Assay you want to analyse</li>
<li>Pick the type of analysis you want to do</li>
<li>Find a package that does it</li>
<li>Find benchmarking papers to narrow the list of packages down</li>
<li>Find the vignette in on the package page</li>
</ul>
</section>
<section id="using-packages-outside-of-intended-usage"
class="slide level2">
<h2>Using Packages Outside of Intended Usage</h2>
<ul>
<li>Do NOT do this</li>
<li>If you absolutely have to, make sure you can justify this to
reviewers before you waste your time</li>
</ul>
</section></section>
<section>
<section id="additional-resources" class="title-slide slide level1">
<h1>Additional Resources</h1>

</section>
<section id="r" class="slide level2">
<h2>R</h2>
<ul>
<li><p><a
href="https://bookdown.org/yihui/rmarkdown/how-to-read-this-book.html">R
Markdown: The Definitive Guide</a> : Excellent R markdown
reference</p></li>
<li><p><a href="https://r4ds.hadley.nz/">R for Data Science</a></p></li>
<li><p><a href="https://ggplot2-book.org/">ggplot2: elegant graphics for
data analysis</a></p></li>
<li><p><a href="https://adv-r.hadley.nz/">Advanced R</a></p></li>
</ul>
</section>
<section id="statistics" class="slide level2">
<h2>Statistics</h2>
<ul>
<li><a href="https://bookdown.org/steve_midway/DAR">Data Analysis in
R</a> : This book has more statistics details than <em>R for Data
Science</em></li>
<li><a
href="https://bookdown.org/steve_midway/DAR/glms-generalized-linear-models.html">Generalized
Linear Models</a><br />
</li>
<li><a
href="https://bookdown.org/steve_midway/DAR/random-effects.html">Random
Effects</a></li>
</ul>
</section>
<section id="rna-seq-analysis" class="slide level2">
<h2>RNA-seq Analysis</h2>
<ul>
<li><a href="https://rnaseq.uoregon.edu/">RNA-seqlopedia</a> :
Everything you need to know about RNA-seq experiments</li>
<li><a
href="https://luisvalesilva.com/datasimple/rna-seq_units.html">RNA-seq
Expression Units</a> : Blog post on understanding common units</li>
<li><a
href="https://bioconductor.org/books/3.17/OSCA.intro/index.html">Introduction
to Single-Cell Analysis with Bioconductor</a> : Covers the basics of
scRNA-seq analysis in R</li>
</ul>
</section>
<section id="dimensional-reduction" class="slide level2">
<h2>Dimensional Reduction</h2>
<ul>
<li><a
href="https://uw.pressbooks.pub/appliedmultivariatestatistics/chapter/pca/">Tutorial
on PCA</a> : PCA explained with R code examples</li>
<li><a
href="https://pair-code.github.io/understanding-umap/">Understanding
UMAP</a> : Short explanation with great visualizations, mainly useful
for scRNA-seq analysis</li>
</ul>
</section></section>
<section id="references" class="title-slide slide level1">
<h1>References</h1>
<ul>
<li>Linear models section was adapted from:
<ul>
<li><a
href="https://ucdavis-bioinformatics-training.github.io/2019-March-Bioinformatics-Prerequisites/thursday/linear_models.html">this
page</a> created by the UC Davis Bioinformatics Core</li>
</ul></li>
</ul>
</section>
    </div>
  </div>

  <script src="C3_bonus_day_files/reveal.js-3.3.0.1/lib/js/head.min.js"></script>
  <script src="C3_bonus_day_files/reveal.js-3.3.0.1/js/reveal.js"></script>

  <script>

      // Full list of configuration options available at:
      // https://github.com/hakimel/reveal.js#configuration
      Reveal.initialize({
        // Display controls in the bottom right corner
        controls: true,
        // Display a presentation progress bar
        progress: true,
        // Push each slide change to the browser history
        history: true,
        // Enable keyboard shortcuts for navigation
        keyboard: true,
        // Enable the slide overview mode
        overview: true,
        // Vertical centering of slides
        center: false,
        // Enables touch navigation on devices with touch input
        touch: true,
        // Turns fragments on and off globally
        fragments: true,
        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,
        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,
        // Stop auto-sliding after user input
        autoSlideStoppable: true,
        // Transition style
        transition: 'default', // none/fade/slide/convex/concave/zoom
        // Transition speed
        transitionSpeed: 'default', // default/fast/slow
        // Transition style for full page slide backgrounds
        backgroundTransition: 'default', // none/fade/slide/convex/concave/zoom
        // Number of slides away from the current that are visible
        viewDistance: 3,

        menu: {
   
    
    
    
    
    
    
 
          custom: false,
          themes: false,
          transitions: false
        },



        // Optional reveal.js plugins
        dependencies: [
          { src: 'C3_bonus_day_files/reveal.js-3.3.0.1/plugin/menu/menu.js', async: true },
        ]
      });
    </script>
  <!-- dynamically load mathjax for compatibility with self-contained -->
  <script>
    (function () {
      var script = document.createElement("script");
      script.type = "text/javascript";
      script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
      document.getElementsByTagName("head")[0].appendChild(script);
    })();
  </script>

<script>
  (function() {
    if (window.jQuery) {
      Reveal.addEventListener( 'slidechanged', function(event) {  
        window.jQuery(event.previousSlide).trigger('hidden');
        window.jQuery(event.currentSlide).trigger('shown');
      });
    }
  })();
</script>


  </body>
</html>
